### ISSUE TITLE
Service worker MessageEvents set the "origin" property, but Worker MessageEvents do not.

### ISSUE THREAD

**Issue by domenic** [w3c/ServiceWorker#1122]
Compare https://html.spec.whatwg.org/#dom-messageport-postmessage and https://w3c.github.io/ServiceWorker/#dom-serviceworker-postmessage . `source` is special in service workers, fine, but I'm confused as to why the inconsistency in `origin`.

I think the best fix might be to set `origin` for workers too, but I am not sure.

**Comment by wanderview** [comment]
FWIW we do initialize the origin in all cases, AFAICT.

**Comment by domenic** [comment]
Great, OK, we should write tests and update HTML probably.

**Comment by wanderview** [comment]
Well, we may not be consistent in doing that.  Our InitMessageEvent() takes an origin, but we do pass empty string in some places.  For example, firing a MessageEvent in a worker due to worker.postMessage() seems to use the empty string origin. :-\

MessageChannel also seems to pass the empty string.

Sorry for my confusion here.  Maybe @bakulf can shed some light on it.

**Comment by wanderview** [comment]
MessagePort code here:

https://dxr.mozilla.org/mozilla-central/source/dom/messagechannel/MessagePort.cpp#155

That first `EmptyString()` is for the origin.

**Comment by domenic** [comment]
Yeah, OK, so Gecko is probably following the current spec, which IMO is pretty inconsistent. I'm not sure why HTML passes no origin in so many places. I'd love to hear if there are any implementation issues in Gecko that would make passing the origin tricky (and if so, help explain why the spec doesn't do it), or if it'd be a straightforward change to start passing an origin everywhere.

/cc @mattto as well for a Blink perspective.

**Comment by jungkees** [comment]
I think conceptually the client's origin would be better than the empty string. Right. Unless there have been implementation issues.

**Comment by annevk** [comment]
I suspect the specification doesn't do it is because we wanted to encourage the concept of object-based capabilities rather than stack inspection. The latter often results in subtle bugs.

**Comment by jungkees** [comment]
> wanted to encourage the concept of object-based capabilities rather than stack inspection. The latter often results in subtle bugs.

@annevk, is there any pointer I can learn more on this?

**Comment by annevk** [comment]
https://html.spec.whatwg.org/#ports-as-the-basis-of-an-object-capability-model-on-the-web

**Comment by yoshisatoyanagisawa** [comment]
Triage: since ServiceWorker are also in the same origin, having origin might not be necessary, but removing that from the spec should be the other issue.  Let me mark this not planned.

**Comment by annevk** [comment]
What's the other issue?

**Comment by asutherland** [comment]
Replying to https://github.com/w3c/ServiceWorker/issues/1122#issuecomment-296311526
> I'd love to hear if there are any implementation issues in Gecko that would make passing the origin tricky (and if so, help explain why the spec doesn't do it), or if it'd be a straightforward change to start passing an origin everywhere.

This would be a straightforward change in gecko.  [Here is where we pass an empty origin for to-parent messages](https://searchfox.org/firefox-main/rev/28398d11d39984cc5db08020671e7dd0d1408ec0/dom/workers/MessageEventRunnable.cpp#167-170) and [here is where we pass an empty origin for to-worker messages](https://searchfox.org/firefox-main/rev/28398d11d39984cc5db08020671e7dd0d1408ec0/dom/workers/MessageEventRunnable.cpp#72-75).  I have not done the archeology but presumably the assumption was that since workers are required to be same origin it was silly to pass it; the existence of web extension content scripts has of course clouded that situation slightly.

**Comment by annevk** [comment]
cc @mikewest

**Comment by yoshisatoyanagisawa** [comment]
The editor's triage in https://github.com/w3c/ServiceWorker/issues/1122#issuecomment-3470764261 went to the opposite direction to https://github.com/w3c/webappsec/blob/main/meetings/2025/2025-11-TPAC-minutes.md#postmessageorigin, which @mikewest proposed in the last TPAC.
I also would like to know Mike's thought.
